File: python/section1/q05.py
def make_ngram(n=2, sequence=None):
    if None == sequence:
        print('E: Empty sequence.')
        return None
    else:
        seqlist = []
        for i in range(n):
            seqlist.append(sequence[i:])
        ngram = [[seq[j] for seq in seqlist] for j in range(len(seqlist[-1]))]
        return ngram

if __name__=='__main__':
    string = "I am an NLPer"
    word_bi_gram = make_ngram(sequence=string.split(' '))
    char_bi_gram = make_ngram(sequence=string)
    print(word_bi_gram)
    print(char_bi_gram)

File: scala/section1/src/q05.scala
/**
  * Created by nomotoeriko on 2017/04/20.
  */
object q05 {
  def main(args: Array[String]): Unit = {
    val sentence = "I am an NLPer"
    make_ngram(2, sentence.split(' ').toList).foreach(x => print("(%s)".format(x.mkString(", "))))
    println()
    make_ngram(2, sentence.toList).foreach(x => print("(%s)".format(x.mkString(", "))))
  }
  def make_ngram(n: Int, seq: List[Any]): List[List[Any]]={
    var seqs = List.empty[List[Any]]
    for (i <- 0 until n) seqs  = seqs :+ seq.drop(i)
    var grams = List.empty[List[Any]]
    for (i <- seqs.last.indices) grams = grams :+ seqs.map(x => x(i))
    grams
  }
}
